### PR TITLE
Ticket/6685

### PR DESCRIPTION
**Pull request by xeno-by**

### CODE REVIEW

**Comment by xeno-by** [comment]
Tests will follow shortly

**Comment by xeno-by** [comment]
review @paulp @hubertp

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -4521,7 +4521,8 @@ trait Typers extends Modes with Adaptations with Tags {
                   if (tag.isEmpty) MissingClassTagError(tree, tagType)
                   else new ApplyToImplicitArgs(Select(tag, nme.newArray), args)
                 }
-                typed(newArrayApp, mode, pt)
+                if (newArrayApp.isErrorTyped) newArrayApp
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I haven't debugged this thing but where is the error coming from? If this is from resolveClassTag then tag should be empty, if the MissingClassTagError is a problem then I don't think newArrayApp needs to be created at all.

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: This thing is coming from line 4521 to report an error about the fact that resolveClassTag returned an empty tree.

Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Ok, so why don't you just return the 'tree' there?

Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I mean not 'return tree' but '{ MissingClassTagError(tree, tagType); tree }' ?

Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Actually looking at MissingClassTagError this is not even necessary, just moving typed to ApplyToImplicitArgs should be enough.

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: But that's exactly what's done by https://github.com/scalamacros/kepler/blob/797f4ac97087de5a59ed2f765fa9afb1757bbfba/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala#L611, right?

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Done

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by xeno-by** [comment]
There's a curious difference in behaviour between the normal invocation of scalac (#1) and the specially crafted environment (#2, described in JIRA).

Before going into the difference, I'd like to expose a horrible truth about the macro engine. Currently it catches all throwables coming form macro invocations, wraps them and converts them in type errors. By all throwables I mean really all, including stack overflows.

In #1 the engine can actually catch stack overflows originating from unbounded recursion in Typers and report them as errors. But because there's already an error at the position of those reports, it masks the subsequent SO reports, and the end result looks completely innocent.

To the contrast, for the reason unknown, in #2 stackoverflows evade the catch-all clause in Macros and manifest themselves as compiler crashes. Probably, this has to something with a custom classloader used to run the compiler.

Bottom line: 1) this PR fixes the root of the bug, 2) the bug was undetected until now, because of a design issue exception handling in macros, 3) I will fix that problem in a later pull request to 2.10.x, 4) because of 2-3 there's imho not much sense in writing a test case for this PR, especially taking into account the complicated environment setup procedure.

**Comment by retronym** [comment]
> To the contrast, for the reason unknown, in #2 stackoverflows evade the catch-all clause in Macros and manifest themselves as compiler crashes.

The problems arise from code that ends up catching `StackOverflowError` before the stack has had a chance to unwind, and triggering a follow-up SOE in the exception handler. This can also happen a shade earlier if LHS of the exception handler pattern triggers classloading, e.g. of `NonFatal$`. So it seems the environmental difference between SBT and vanilla scalac is either which classes are already loaded before it gets to that point, or how much stack the classloader needs to do its business.

To review the exception handling, we chould induce a SOE (perhaps just in a macro body), and check that even under vanilla scalac that this is reported verbatim and that it's not treated as a type error with a fallback.

We should also be aware that `scala.util.control.NonFatal` intentionally catches SOE. But we see that it's not always straightforward to recover from it as I had thought.

**Comment by retronym** [comment]
This patch lgtm for 2.10.0.

**Comment by xeno-by** [comment]
@retronym I've opened a bug for the macro exception handling part: https://issues.scala-lang.org/browse/SI-6748.

**Comment by xeno-by** [comment]
@adriaanm @gkossakowski @jsuereth please, merge - this is a blocker for 2.10.0

**Comment by hubertp** [comment]
lgtm

**Comment by xeno-by** [comment]
2 lgtms, >2 days, merging at will

**Comment by adriaanm** [comment]
@xeno-by, thanks for keeping things moving

one of the reasons i like to merge myself is that i make sure the milestone field is set when i merge so we can later more easily get statistics out of github -- I'm working on a doc (and ideally, automation) for this

until then, when merging, could you make sure the milestone is set? thanks!

**Comment by xeno-by** [comment]
Oh I'm sorry didn't know about milestones. Sure I will be more thorough in the future. Is there something else I need to know?
